1. Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 

In [1]:
import pandas as pd

import requests
from urllib.parse import urlencode

### Описание данных

Hotel – тип отеля (City Hotel или Resort Hotel)  

Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным

Lead time – количество дней, прошедших между датой бронирования и датой прибытия  

Arrival full date – полная дата прибытия

Arrival date year – год прибытия  

Arrival date month – месяц прибытия  

Arrival date week number – номер недели прибытия

Arrival date day of month – день прибытия

Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)

Adults – число взрослых

Children – число детей

Babies – число младенцев 

Meal – выбранный тип питания

Country – страна происхождения клиента

Reserved room type – тип зарезервированного номера

Assigned room type – тип полученного номера (может отличаться от забронированного)

Customer type – тип бронирования

Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину
Reservation status date – дата обновления статуса

In [2]:
# Ссылка с необходимыми данными
base_url = 'https://cloud-api.yandex.net/v1/disk/public/resources/download?'
public_key = 'https://disk.yandex.ru/d/OzaUNarkTYbF_Q'

# Получаем загрузочную ссылку
final_url = base_url + urlencode(dict(public_key=public_key))
response = requests.get(final_url)
download_url = response.json()['href']

In [3]:
# Загружаем данные
bookings = pd.read_csv(download_url, sep = ';')
bookings.head(7)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

In [4]:
bookings.shape

(119390, 21)

In [5]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

2. Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.

In [6]:
# Используем list comprehension для измнения регистра букв и пробелов между слов
python_columns = [word.lower().replace(' ', '_') for word in bookings.columns]
bookings = bookings.rename(columns = dict(zip(bookings.columns, python_columns)))

In [7]:
print(bookings.columns)

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')


3. Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.

In [8]:
bookings.query("is_canceled == 0").groupby('country').is_canceled.value_counts().sort_values(ascending = False).head(5)

country  is_canceled
PRT      0              21071
GBR      0               9676
FRA      0               8481
ESP      0               6391
DEU      0               6069
Name: is_canceled, dtype: int64

4. На сколько ночей (stays_total_nights)  в среднем бронируют отели типа City Hotel? Resort Hotel? 

In [9]:
bookings.groupby('hotel').agg({'stays_total_nights': 'mean'})

stays_total_nights
hotel                           
City Hotel              2.978142
Resort Hotel            4.318547

5. Иногда тип номера, присвоенного клиенту (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга.

Сколько подобных наблюдений встретилось в датасете?

*отмена бронирования также считается

In [10]:
bookings.query('reserved_room_type != assigned_room_type').shape[0]

14917

6. Теперь проанализируйте даты запланированного прибытия (arrival_date_year).

На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?

In [11]:
bookings \
        .groupby('arrival_date_year') \
        .arrival_date_month \
        .value_counts() \
        .sort_values(ascending = False)

arrival_date_year  arrival_date_month
2017               May                   6313
2016               October               6203
2017               April                 5661
                   June                  5647
2016               May                   5478
                   April                 5428
                   September             5394
2017               July                  5313
2016               June                  5292
2015               September             5114
2016               August                5063
2017               March                 4970
2015               October               4957
2017               August                4925
2016               March                 4824
                   July                  4572
                   November              4454
2017               February              4177
2016               February              3891
2015               August                3889
2016               December              3

В 2017 чаще всего оформляли бронирование на май  
В 2016 чаще всего бронировали отель на октябрь

7. Сгруппируйте данные по годам, а затем проверьте, на какой месяц (arrival_date_month) бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017? 

In [12]:
bookings \
    .query('is_canceled == 1') \
    .groupby(['arrival_date_year', 'arrival_date_month'])['is_canceled'] \
    .value_counts() \
    .sort_values(ascending = False)

arrival_date_year  arrival_date_month  is_canceled
2017               May                 1              2762
2016               October             1              2514
2017               April               1              2463
                   June                1              2439
2016               June                1              2096
2015               September           1              2094
2016               April               1              2061
                   September           1              2022
2017               July                1              1984
2016               May                 1              1915
                   August              1              1825
2017               August              1              1816
2015               October             1              1732
2017               March               1              1672
2016               November            1              1636
2015               August              1              1598
2016 

Чаще всего бронирования отменялись:  
2015 - сентябрь  
2016 - октябрь  
2017 - Май  



8. Посмотрите на числовые характеристики трёх колонок: adults, children и babies. Какая из них имеет наибольшее среднее значение?

In [13]:
bookings. \
    agg(dict(zip(['adults', 'children','babies'], ['mean', 'mean', 'mean'])))

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

В среднем больше отель бронируют с подростками, среднее значение: 1.856403

9. Создайте колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?

In [14]:
bookings['total_kids'] = bookings.children + bookings.babies

In [15]:
bookings. \
    groupby('hotel'). \
    agg({'total_kids': 'mean'}).round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

10. Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитать метрику под названием Churn Rate.  
В нашем случае Churn Rate - это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика зависит от наличия детей у клиентов!

In [16]:
bookings['has_kids'] = bookings.total_kids != 0

In [17]:
# Количество бронирований без детей
without_kids = bookings. \
        query('has_kids == False'). \
        shape[0]
print(without_kids)

110054


In [18]:
# Количетсво отмененных бронирований без детей
without_kids_c = bookings. \
        query('has_kids == False and is_canceled == 1'). \
        shape[0]
print(without_kids_c)

40961


In [19]:
churn_rate_wc = without_kids_c / without_kids
print(round(churn_rate_wc * 100, 2))

37.22


In [20]:
# Количетсво бронирований с детьми
with_kids = bookings. \
        query('has_kids == True'). \
        shape[0]
print(with_kids)

9336


In [21]:
# Количетсво отмененных бронирований с детьми
with_kids_c = bookings. \
        query('has_kids == True and is_canceled == 1'). \
        shape[0]
print(with_kids_c)

3263


In [22]:
churn_rate_c = with_kids_c / with_kids
print(round(churn_rate_c * 100, 2))

34.95


Клиенты без детей чаще отменяли бронь, churn rate составляет 37.22